File: convert_to_markdown.py
#!/usr/bin/env python3
import os
import re
from bs4 import BeautifulSoup
import html2text
from pathlib import Path

def convert_html_to_markdown(html_file):
    with open(html_file, 'r', encoding='utf-8') as f:
        content = f.read()
    
    soup = BeautifulSoup(content, 'html.parser')
    
    # Extract main content
    main_content = soup.find('main', class_='wrapper text')
    if not main_content:
        return None
    
    # Initialize markdown content
    md_content = []
    
    # Add chapter number if exists
    chapter_div = main_content.find('div', class_='chapter')
    if chapter_div:
        md_content.append(f"# {chapter_div.text.strip()}\n")
    
    # Add title
    title = main_content.find('h1')
    if title:
        md_content.append(f"# {title.text.strip()}\n")
    
    # Process paragraphs and other elements
    for element in main_content.find_all(['p', 'h2', 'h3', 'h4', 'figure', 'pre']):
        if element.name == 'p':
            # Handle paragraphs with RTL text
            text = element.get_text().strip()
            if text:
                md_content.append(f"<div dir='rtl'>{text}</div>\n")
        
        elif element.name in ['h2', 'h3', 'h4']:
            # Handle headings with RTL text
            level = int(element.name[1])
            text = element.get_text().strip()
            if text:
                md_content.append(f"{'#' * level} {text}\n")
        
        elif element.name == 'figure':
            # Handle images
            img = element.find('img')
            if img:
                src = img.get('src', '')
                alt = img.get('alt', '')
                md_content.append(f"![{alt}]({src})\n")
            
            # Handle code blocks
            if 'class' in element.attrs and 'code' in element['class']:
                figcaption = element.find('figcaption')
                if figcaption:
                    md_content.append(f"<div dir='ltr'>\n")
                    md_content.append(f"**{figcaption.text.strip()}**\n")
                    pre = element.find('pre')
                    if pre:
                        md_content.append("```\n")
                        md_content.append(pre.text)
                        md_content.append("\n```\n")
                    md_content.append("</div>\n")
    
    return '\n'.join(md_content)

def process_directory():
    # Create markdown directory if it doesn't exist
    markdown_dir = Path('markdown')
    markdown_dir.mkdir(exist_ok=True)
    
    # Process all HTML files
    for html_file in Path('.').glob('*.html'):
        if html_file.name.startswith('.'):
            continue
            
        print(f"Converting {html_file.name}...")
        md_content = convert_html_to_markdown(html_file)
        
        if md_content:
            md_file = markdown_dir / f"{html_file.stem}.md"
            with open(md_file, 'w', encoding='utf-8') as f:
                f.write(md_content)

if __name__ == '__main__':
    process_directory() 

File: convert_to_markdown.js
const fs = require('fs');
const path = require('path');
const cheerio = require('cheerio');

function convertHtmlToMarkdown(htmlFile) {
    const content = fs.readFileSync(htmlFile, 'utf8');
    const $ = cheerio.load(content);
    
    // Extract main content
    const mainContent = $('main.wrapper.text');
    if (!mainContent.length) return null;
    
    const mdContent = [];
    
    // Add chapter number if exists
    const chapterDiv = mainContent.find('div.chapter');
    if (chapterDiv.length) {
        mdContent.push(`# ${chapterDiv.text().trim()}\n`);
    }
    
    // Add title
    const title = mainContent.find('h1');
    if (title.length) {
        mdContent.push(`# ${title.text().trim()}\n`);
    }
    
    // Process paragraphs and other elements
    mainContent.find('p, h2, h3, h4, figure, pre').each((_, element) => {
        const $element = $(element);
        
        if (element.tagName === 'p') {
            const text = $element.text().trim();
            if (text) {
                mdContent.push(`<div dir='rtl'>${text}</div>\n`);
            }
        }
        else if (['h2', 'h3', 'h4'].includes(element.tagName)) {
            const level = parseInt(element.tagName[1]);
            const text = $element.text().trim();
            if (text) {
                mdContent.push(`${'#'.repeat(level)} ${text}\n`);
            }
        }
        else if (element.tagName === 'figure') {
            // Handle images
            const img = $element.find('img');
            if (img.length) {
                const src = img.attr('src') || '';
                const alt = img.attr('alt') || '';
                mdContent.push(`![${alt}](${src})\n`);
            }
            
            // Handle code blocks
            if ($element.hasClass('code')) {
                const figcaption = $element.find('figcaption');
                if (figcaption.length) {
                    mdContent.push(`<div dir='ltr'>\n`);
                    mdContent.push(`**${figcaption.text().trim()}**\n`);
                    const pre = $element.find('pre');
                    if (pre.length) {
                        mdContent.push('```\n');
                        mdContent.push(pre.text());
                        mdContent.push('\n```\n');
                    }
                    mdContent.push('</div>\n');
                }
            }
        }
    });
    
    return mdContent.join('\n');
}

function processDirectory() {
    // Create markdown directory if it doesn't exist
    const markdownDir = path.join(__dirname, 'markdown');
    if (!fs.existsSync(markdownDir)) {
        fs.mkdirSync(markdownDir);
    }
    
    // Process all HTML files
    fs.readdirSync(__dirname).forEach(file => {
        if (file.endsWith('.html') && !file.startsWith('.')) {
            console.log(`Converting ${file}...`);
            const mdContent = convertHtmlToMarkdown(path.join(__dirname, file));
            
            if (mdContent) {
                const mdFile = path.join(markdownDir, file.replace('.html', '.md'));
                fs.writeFileSync(mdFile, mdContent, 'utf8');
            }
        }
    });
}

processDirectory(); 